### PR TITLE
Aoc practice

### PR DESCRIPTION
**Pull request by LucyxxNie**

### CODE REVIEW

File: src/aoc2020/day1.clj [hunk 1]
```diff
@@ -1,61 +1,75 @@
...
-
-                )))
-       (range 0 (-> data-set count))
-       (range (-> data-set count) 0 -1))
-  )
+  (:require [aoc2020.util :refer [file->seq]]))
+
+(defn parse-int
+  [entry-str]
+  (Integer/parseInt entry-str))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: Built-in version of `parse-long` will nil pun bad values

File: src/aoc2020/day1.clj [hunk 2]
```diff
@@ -1,61 +1,75 @@
...
+  [sum x]
+  (- sum x))
+
+(def complement-2020
+  (partial complement 2020))
+
+(defn two-sum
+  [data]
+  (let [set                 (into #{} data)
+        compl-exist?        (fn [x] (contains? set (complement-2020 x)))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: This is misnamed (for what it is). The name should likely remain the same, and should take two args.

File: src/aoc2020/day1.clj [hunk 3]
```diff
@@ -1,61 +1,75 @@
...
+        result              (filterv compl-exist? set)]
+
+    (* (result 0) (result 1))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: Use explicit retrieval and give them a name (in the let block).

File: src/aoc2020/day1.clj [hunk 4]
```diff
@@ -1,61 +1,75 @@
...
+    ))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: The fns shouldn't have any trailing parens

File: src/aoc2020/day1.clj [hunk 5]
```diff
@@ -1,61 +1,75 @@
...
-  (->> map (fn [left-pointer
-                right-pointer]
-             ((if (== 2020 (+ (get data-set left-pointer)
-                              (get data-set right-pointer)))
-
-                )))
-       (range 0 (-> data-set count))
-       (range (-> data-set count) 0 -1))
-  )
+  (:require [aoc2020.util :refer [file->seq]]))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: Use IntelliJ/Cursive's autoindent for any files before you check in code

File: src/aoc2020/day1.clj [hunk 6]
```diff
@@ -1,61 +1,75 @@
...
+  [data]
+  (let [set                 (into #{} data)
+        compl-exist?        (fn [x] (contains? set (complement-2020 x)))
+        result              (filterv compl-exist? set)]
+
+    (* (result 0) (result 1))
+    ))
+
+
+(defn three-sum-step1
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: misnamed

File: src/aoc2020/day1.clj [hunk 7]
```diff
@@ -1,61 +1,75 @@
...
+  [data num]
+  (let [set                 (into #{} data)
+        compl               (complement-2020 num)
+        compl-exist?        (fn [x] (contains? set (complement compl x)))
+        result              (filterv compl-exist? set)]
+
+    (if (first result)
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: See `when`

File: src/aoc2020/day1.clj [hunk 8]
```diff
@@ -1,61 +1,75 @@
...
+(defn three-sum-step2
+  [data]
+  (->> (mapv
+         (fn [x] (three-sum-step1 data x))
+         data)
+       (filterv (fn [n] n))
+       (first)))
 
 
 (comment
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: This Rich Comment Block is **bare**.  The evaluation of the functions and how they work should be the _first_ thing developed, not the last. (In this case it doesn't really exist). 

This is likely the root cause for many of the issues above

File: src/aoc2020/day1.clj (hunk 2 above)
Comment (LucyxxNie) [reply src/aoc2020/day1.clj]: Are you referring to two-sum or complement-2020?

File: src/aoc2020/day1.clj (hunk 5 above)
Comment (LucyxxNie) [reply src/aoc2020/day1.clj]: I have used the default setting which is 2 spaces indentation from IntelliJ/cursive. Please let me know if we need to adjust it.

Comment (lukaszhall) [reply src/aoc2020/day1.clj]: Look up the auto-indent hotkeys in IntelliJ _**and use them**_. Most of the binding forms are not currently indented properly

File: src/aoc2020/day1.clj [hunk 9]
```diff
@@ -1,61 +1,102 @@
...
-       (range (-> data-set count) 0 -1))
-  )
+  (:require [aoc2020.util :refer [file->seq]]))
+
+(defn complement
+  [sum x]
+  (- sum x))
+
+
+(defn find-complement
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: avoid defaulting to verb naming. What is returned here? Based purely on the names, how would someone know the difference between `complement` and `find-complement` ?

File: src/aoc2020/day1.clj [hunk 10]
```diff
@@ -1,61 +1,102 @@
...
+  [data-set sum]
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: inverted args

File: src/aoc2020/day1.clj [hunk 11]
```diff
@@ -1,61 +1,102 @@
...
+  (let [compl-exist?  (fn [num]
+                        (contains? data-set (complement sum num)))
+        result        (filterv compl-exist? data-set)]
+
+    (when (first result)
+      result)))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: `first` of an empty collection is already `nil`

File: src/aoc2020/day1.clj [hunk 12]
```diff
@@ -1,61 +1,102 @@
...
+
+(defn two-sum-product
+  [data-s sum]
+  (let [data-set     (into #{} data-s)
+        result       (find-complement data-set sum)
+        num1         (result 0)
+        num2         (result 1)]
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: This was already addressed in a previous PR... using dynamically bound symbols as functions is dangerous and difficult to debug

File: src/aoc2020/day1.clj [hunk 13]
```diff
@@ -1,61 +1,102 @@
...
+    (* num1 num2)))
+
+
+(defn compl-product
+  [data-s sum num]
+  (let [data-set     (into #{} data-s)
+        compl        (complement sum num)
+        result       (find-complement data-set compl)
+        num2         (get result 0)
+        num3         (get result 1)]
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: See `take`

File: src/aoc2020/day1.clj [hunk 14]
```diff
@@ -1,61 +1,102 @@
...
+
+  (do (def entries (->> (file->seq "aoc2020/day1/input.txt")
+                        (mapv parse-long)))
+      entries)
+  #_=> [1348
+        1621
+        1500
+        1818
+        1266
+        1449,,,]
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: These belong at the top of the RCB. There are times when you'll want to demonstrate the functions against simple data sets, and other times when you'll want to demonstrate those functions as sample data.


---
EDIT: These are used above. You _can not_ use a symbol before it is defined (unless you `declare` it, but that's not what you're attempting to do here)

File: src/aoc2020/day1.clj [hunk 15]
```diff
@@ -1,61 +1,102 @@
...
+
+(defn complement
+  [sum x]
+  (- sum x))
+
+
+(defn find-complement
+  [data-set sum]
+  (let [compl-exist?  (fn [num]
+                        (contains? data-set (complement sum num)))
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: Seems like this function (`complement?`) should be defined externally?  It's very rare that you will want to define anonymous functions and name them, but hid them within another function definition

File: src/aoc2020/day1.clj [hunk 16]
```diff
@@ -1,61 +1,102 @@
...
+;------------------------------data----------------------------------------
+
+  (do (def sample-entry-s (->> (file->seq "aoc2020/day1/input-sample.txt")
+                               (mapv parse-long)))
+      sample-entry-s)
+  #_=> [1721 979 366 299 675 1456]
+
+  (do (def entries (->> (file->seq "aoc2020/day1/input.txt")
+                        (mapv parse-long)))
+      entries)
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: shouldn't this be `entry-s`?

File: src/aoc2020/day1.clj [hunk 17]
```diff
@@ -1,61 +1,102 @@
...
-    988
-    ...)
+  ;;------------------------function evaluation---------------------------
+
+  (complement 2020 20)
+  #_=> 2000
+  (mapv (partial complement 2020) [1 6 5 23 90 3])
+  #_=> [2019 2014 2015 1997 1930 2017]
+  (mapv (partial complement 2020) sample-entry-s)
+  #_=> [299 1041 1654 1721 1345 564]
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: These feel like copy-paste examples.   i.e., the third is not particularly useful to demonstrate the function, while the second is much more useful and self-contained. The first example is just a duplicate subset of the second... why is it still there? Is 2020 the best sum to illustrate this example? Are `[1 6 5 23 90 3]` the examples you would given to demonstrate how something works, as well as edge cases?

Same idea applies below

File: src/aoc2020/day1.clj [hunk 18]
```diff
@@ -1,62 +1,113 @@
...
-    695
-    934
-    974
-    988
-    ...)
+  ;------------------------------data----------------------------------------
+
+  (do (def sample-entry-s (->> (file->seq "aoc2020/day1/input-sample.txt")
+                            (mapv parse-long)))
+    sample-entry-s)
```
Comment (lukaszhall) [inline src/aoc2020/day1.clj]: There's something wrong with your indentation. Check in with Leo (we review it yesterday) to make sure you are using your indent hotkey.

**Review by lukaszhall — CHANGES_REQUESTED** [review]
The biggest remaining gaps here are vague naming (num? x?, ...etc) and lack of thorough decomposition.

For example, there's a large gap between two-sum and three-sum (in the RCB).

**Pull request merged**